### PR TITLE
Consider variable names of fused function when looking for the loop min/max in compute_shift_factor

### PR DESCRIPTION
**Pull request by vksnk**
This fixes the issue from #4691. As described there, fused var names may not match var names in the given function, so search in the list fails. This PR adds an extra check for the case, when we can't find the name and tries to search under different name of fused var.

### CODE REVIEW

File: src/ScheduleFunctions.cpp [hunk 1]
```diff
@@ -1193,10 +1193,17 @@ class InjectFunctionRealization : public IRMutator {
             internal_assert(iter != dims.end());
             start_fuse = (int)(iter - dims.begin());
         }
+
+        const auto &env_iter = env.find(fuse_level.func());
+        internal_assert(env_iter != env.end());
+        const auto &parent_func = env_iter->second;
+
+        const vector<Dim> &parent_dims = (fuse_level.stage_index() == 0) ? parent_func.definition().schedule().dims() : parent_func.update(fuse_level.stage_index() - 1).schedule().dims();
```
Comment (steven-johnson) [inline src/ScheduleFunctions.cpp]: style nit: long line should be broken, eg

```
const vector<Dim> &parent_dims = (fuse_level.stage_index() == 0) 
    ? parent_func.definition().schedule().dims() 
    : parent_func.update(fuse_level.stage_index() - 1).schedule().dims();
```

File: src/ScheduleFunctions.cpp [hunk 2]
```diff
@@ -1193,10 +1193,17 @@ class InjectFunctionRealization : public IRMutator {
...
+        int fused_vars_num = dims.size() - start_fuse - 1;
```
Comment (steven-johnson) [inline src/ScheduleFunctions.cpp]: Does the code above guarantee this won't underflow?

File: src/ScheduleFunctions.cpp [hunk 3]
```diff
@@ -1211,17 +1218,27 @@ class InjectFunctionRealization : public IRMutator {
             }
 
             string parent_prefix = fuse_level.func() + ".s" + std::to_string(fuse_level.stage_index()) + ".";
+            int parent_var_index = (i - start_fuse) + (int)parent_dims.size() - 1 - fused_vars_num;
```
Comment (steven-johnson) [inline src/ScheduleFunctions.cpp]: Ditto underflow question. Is it overkill to add `internal_assert(parent_var_index >=0)`?

File: src/ScheduleFunctions.cpp (hunk 1 above)
Comment (vksnk) [reply src/ScheduleFunctions.cpp]: Done, I was hoping that 'make format' would fix it, but apparently not.

File: src/ScheduleFunctions.cpp (hunk 2 above)
Comment (vksnk) [reply src/ScheduleFunctions.cpp]: There is a check above to make sure that start_fuse is in range, however, I moved this line up to make it more visible.

File: src/ScheduleFunctions.cpp (hunk 3 above)
Comment (vksnk) [reply src/ScheduleFunctions.cpp]: Good point, I've added assert.

File: src/ScheduleFunctions.cpp (hunk 1 above)
Comment (vksnk) [reply src/ScheduleFunctions.cpp]: Ugh, now clang-format pre-check complains, is there a way to suppress it?

Comment (abadams) [reply src/ScheduleFunctions.cpp]: I would trying adding parenthesis around the entire expression and then just accept whatever 'make format' decides to do with it.

Comment (vksnk) [reply src/ScheduleFunctions.cpp]: Thanks, 'make format' would just make it one line no matter what I do. I've split it into two lines, one of them is still pretty long, but maybe closer to acceptable.

**Comment by vksnk** [comment]
I've addressed the comments, please take a look.

**Comment by steven-johnson** [comment]
LGTM but I don't know this part of the code well enough to be the best reviewer.

**Comment by steven-johnson** [comment]
ping to @abadams for review

**Comment by abadams** [comment]
I'm not sure I understand what's going on here or why there are multiple possible names. I think the parent func is the one it's fused with? Could you add more comments that explain it?

**Comment by vksnk** [comment]
This is based on the test case from this PR, but say you have two functions (A and B) fused together (`A.compute_with(B, y)`) and A is split like this:

```
.split(y, yo, y, 32 * 2, TailStrategy::RoundUp)
.split(y, y, yi, 2, TailStrategy::RoundUp)
```
while B like this:

`.split(y, yo, y, 32, TailStrategy::RoundUp)`

if you print the names of dimensions:

for A:

```
x.v0
x.x
y.y.yi
y.y.y <- slightly different name
y.yo
```

for B:

```
x.v1
x.x
c
y.y <- slightly different name
y.yo
```

You still can fuse them, because check that fuse dimensions match is done by comparing them [both ways with ends_with](https://github.com/halide/Halide/blob/master/src/ScheduleFunctions.cpp#L54), so it's sort of symmetrical. However, this specific function I've modified in this PR is doing exact string match/look-up of parent's loop var using the name of child's var, so in case like this, it fails to find anything and asserts.

**Comment by abadams** [comment]
So shouldn't it always use the parent var name then, if it's looking for the parent's loop?

**Comment by vksnk** [comment]
> So shouldn't it always use the parent var name then, if it's looking for the parent's loop?

Good point, this should be enough. Let me fix it.

**Comment by vksnk** [comment]
> > So shouldn't it always use the parent var name then, if it's looking for the parent's loop?
> 
> Good point, this should be enough. Let me fix it.

Done, please take a look.

**Pull request merged**